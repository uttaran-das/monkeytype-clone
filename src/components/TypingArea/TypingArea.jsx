import React from 'react'
import PropTypes from 'prop-types'
import Character from "../Character/Character";
import Cursor from "../Cursor/Cursor";

const TypingArea = ({ prompt, userInput }) => {
    const characters = prompt.split('');

    return (
        <div className="words-container" data-testid="typing-area">
            {characters.map((char, index) => {
                const isTyped = index < userInput.length
                let state
                if (isTyped) state = userInput[index] === char ? 'correct' : 'incorrect'
                const isCurrentChar = index === userInput.length

                return (
                    <span key={index}>
                        {isCurrentChar && <Cursor />}
                        <Character char={char} state={state} />
                    </span>
                )
            })}
            {userInput.length >= prompt.length && <Cursor />}
        </div>
    )
}

TypingArea.propTypes = {
    prompt: PropTypes.string.isRequired,
    userInput: PropTypes.string.isRequired,
}

export default TypingArea