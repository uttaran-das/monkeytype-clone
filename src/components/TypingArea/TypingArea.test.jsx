// src/components/TypingArea/TypingArea.test.jsx
import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import TypingArea from './TypingArea';

describe('TypingArea component', () => {
    const prompt = "Test prompt";

    it('renders correctly and matches snapshot', () => {
        const { container } = render(<TypingArea prompt={prompt} userInput="" />);
        expect(container).toMatchSnapshot();
    });

    it('styles characters correctly based on userInput', () => {
        // User typed "Tesk" instead of "Test"
        const userInput = "Tesk";
        render(<TypingArea prompt={prompt} userInput={userInput} />);

        const tElements = screen.getAllByText('t');

        expect(screen.getByText('T')).toHaveClass('correct');
        expect(screen.getByText('e')).toHaveClass('correct');
        expect(screen.getByText('s')).toHaveClass('correct');
        expect(tElements[0]).toHaveClass('incorrect');
        expect(tElements[1]).not.toHaveClass('incorrect');
        expect(tElements[1]).not.toHaveClass('correct');
    });

    it('positions the cursor after the last typed character', () => {
        const userInput = "Te";
        render(<TypingArea prompt={prompt} userInput={userInput} />);

        const cursor = screen.getByTestId('cursor');
        // The cursor's next sibling in the DOM should be the 's' character span
        const sCharElement = screen.getByText('s');
        expect(cursor.nextSibling).toBe(sCharElement);
    });

    it('positions the cursor at the end if the prompt is fully typed', () => {
        render(<TypingArea prompt={prompt} userInput={prompt} />);
        const cursor = screen.getByTestId('cursor');

        // The cursor should be the last child in its container span
        expect(cursor.nextSibling).toBeNull();
    });
});