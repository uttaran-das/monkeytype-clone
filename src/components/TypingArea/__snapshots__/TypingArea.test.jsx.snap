// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TypingArea component > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="words-container"
    data-testid="typing-area"
  >
    <span>
      <span
        class="cursor"
        data-testid="cursor"
      />
      <span
        class="character "
      >
        T
      </span>
    </span>
    <span>
      <span
        class="character "
      >
        e
      </span>
    </span>
    <span>
      <span
        class="character "
      >
        s
      </span>
    </span>
    <span>
      <span
        class="character "
      >
        t
      </span>
    </span>
    <span>
      <span
        class="character "
      >
        Â 
      </span>
    </span>
    <span>
      <span
        class="character "
      >
        p
      </span>
    </span>
    <span>
      <span
        class="character "
      >
        r
      </span>
    </span>
    <span>
      <span
        class="character "
      >
        o
      </span>
    </span>
    <span>
      <span
        class="character "
      >
        m
      </span>
    </span>
    <span>
      <span
        class="character "
      >
        p
      </span>
    </span>
    <span>
      <span
        class="character "
      >
        t
      </span>
    </span>
  </div>
</div>
`;
